cdap-security/src/test/java/co/cask/cdap/security/impersonation/UGIProviderTest.java

```diff
@@ -140,97 +155,205 @@ public static void finish() {
+                        aliceUGIWithPrincipal.getUGI().getAuthenticationMethod());
+    Assert.assertEquals(aliceKerberosPrincipalId.getPrincipal(), aliceUGIWithPrincipal.getPrincipal());
+    Assert.assertTrue(aliceUGIWithPrincipal.getUGI().hasKerberosCredentials());
 
     // Fetch it again, it is should return the same UGI since there is caching
-    Assert.assertSame(configuredUGI.getUGI(), provider.getConfiguredUGI(opInfo));
-
-//    // Put the keytab on HDFS
-//    Location remoteKeytab = locationFactory.create("keytab").getTempFile(".tmp");
-//    Files.copy(keytabFile, Locations.newOutputSupplier(remoteKeytab));
-//
-//    // Login with remote keytab file
-//    ImpersonationInfo bobInfo = new ImpersonationInfo(getPrincipal("bob"), remoteKeytab.toURI().toString());
-//    UserGroupInformation bobUGI = provider.getConfiguredUGI(bobInfo);
-//    Assert.assertEquals(UserGroupInformation.AuthenticationMethod.KERBEROS, bobUGI.getAuthenticationMethod());
-//    Assert.assertTrue(bobUGI.hasKerberosCredentials());
-//
-//    // Delete the keytab on HDFS
-//    remoteKeytab.delete();
-//
-//    // Fetch the bob UGI again, it should still return the valid one
-//    Assert.assertSame(bobUGI, provider.getConfiguredUGI(bobInfo));
-//
-//    // Invalid the cache, getting of Alice UGI should pass, while getting of Bob should fails
-//    provider.invalidCache();
-//    Assert.assertNotSame(aliceUGI, provider.getConfiguredUGI(aliceInfo));
-//    try {
-//      provider.getConfiguredUGI(bobInfo);
-//      Assert.fail("Expected IOException when getting UGI for " + bobInfo);
-//    } catch (IOException e) {
-//      // Expected
-//    }
+    Assert.assertSame(aliceUGIWithPrincipal.getUGI(), provider.getConfiguredUGI(aliceImpRequest).getUGI());
+
+    ImpersonationRequest bobImpRequest = new ImpersonationRequest(bobEntity, ImpersonatedOpType.OTHER);
+    UGIWithPrincipal bobUGIWithPrincipal = provider.getConfiguredUGI(bobImpRequest);
+    Assert.assertEquals(UserGroupInformation.AuthenticationMethod.KERBEROS,
+                        bobUGIWithPrincipal.getUGI().getAuthenticationMethod());
+    Assert.assertTrue(bobUGIWithPrincipal.getUGI().hasKerberosCredentials());
+    Assert.assertEquals(bobKerberosPrincipalId.getPrincipal(), bobUGIWithPrincipal.getPrincipal());
+
+    Assert.assertTrue(bobKeytabFile.delete());
+
+    // Fetch the bob UGI again, it should still return the valid one
+    Assert.assertSame(bobUGIWithPrincipal, provider.getConfiguredUGI(bobImpRequest));
+
+    // Invalid the cache, getting of Alice UGI should pass, while getting of Bob should fails
+    provider.invalidCache();
+    Assert.assertNotSame(aliceUGIWithPrincipal, provider.getConfiguredUGI(aliceImpRequest));
+    try {
+      provider.getConfiguredUGI(bobImpRequest);
+      Assert.fail("Expected IllegalArgumentException when getting UGI for " + bobUGIWithPrincipal);
+    } catch (IllegalArgumentException e) {
```

Comment:
Is this the expected behavior? If the user does not exist (or the keytab), we throw IllegalArgument? Shouldn't it be a NotFoundException or something like that? I would expect that IllegalArgument indicates that the bob request is ill-formed. Such as invalid characters in entity name. Can you file a Jira and add a TODO for an improvement?